devicetokens/devicetokens.go

```diff
@@ -0,0 +1,105 @@
+package devicetokens
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/tidepool-org/platform/errors"
+	"github.com/tidepool-org/platform/structure"
+)
+
+const (
+	AppleEnvProduction = "production"
+	AppleEnvSandbox    = "sandbox"
+
+	// MaxTokenLen for an opaque token blob sent by Apple.
+	//
+	// Apple's docs indicate that the length should not be hard-coded, but
+	// we've decided this is an appropriate maximum limit. Assuming the blob
+	// is just a randomly generated identifier, there's no forseeable reason
+	// it should require anywhere near this much. There are only 128 bits
+	// (that's a mere 16 bytes!) in a UUID afterall.
+	MaxTokenLen = 8192
+)
+
+type Document struct {
+	// UserID of the user that owns the DeviceToken.
+	UserID string `json:"userId" bson:"userId"`
+	// TokenKey is string that uniquely identifies the DeviceToken.
+	//
+	// It's useful for generating unique indexes.
+	TokenKey string `json:"tokenKey" bson:"tokenKey"`
+	// DeviceToken wraps the device-specific token.
+	DeviceToken DeviceToken `json:"deviceToken" bson:"deviceToken"`
+}
+
+func NewDocument(userID string, deviceToken DeviceToken) *Document {
+	return &Document{
+		UserID:      userID,
+		TokenKey:    deviceToken.key(),
+		DeviceToken: deviceToken,
+	}
+}
+
+// DeviceToken is received from a Tidepool client application.
+//
+// It contains the information necessary for a service to send a push
+// notification to the device.
+type DeviceToken struct {
+	// Apple devices should provide this information.
+	Apple *AppleDeviceToken `json:"apple,omitempty" bson:"apple,omitempty"`
+}
+
+// key provides a unique string value to identify this device token.
+//
+// Intended to be used as part of a unique index for database indexes.
+func (t DeviceToken) key() string {
+	if t.Apple != nil {
+		return t.Apple.key()
+	}
+	return ""
+}
+
+func (t DeviceToken) Validate(validator structure.Validator) {
+	if t.Apple != nil {
+		t.Apple.Validate(validator)
+	} else {
+		// There's no other kind of token, so if there's no Apple, this is invalid.
+		validator.ReportError(errors.New("no token found"))
+	}
+}
+
+type AppleDeviceToken struct {
+	// Token from Apple that identifies this specific device.
+	Token AppleBlob
+	// Environment is either sandbox or production.
+	Environment string
+}
```

Comment:
FWIW- bson automatically does that[^1], and Go by default will look for Token first, but then look for token, so either can be used[^2]. Adding tags here is just noise. 🤷🏻 

[1]: https://pkg.go.dev/go.mongodb.org/mongo-driver/bson#hdr-Structs
[2]: https://pkg.go.dev/encoding/json#Unmarshal

Reply:
Understood. I disagree that it is noise. To me, it is being very explicit about what is expected and what is not. Also, it makes it clear that it is expected to be used for both JSON and BSON (because it has both tags). Whether you agree with that or not, that is convention in the platform code.

Reply:
In either event, I've made the change that you asked for. :)

If you want to think about it more though, let me challenge your thinking that it's more explicit. Consider that you can have `json:"token"`, and pass it "Token":"blah", and it will still be accepted. That's how the Go JSON parser works. You tried to be more explicit with the tag, but if anything, it's just misleading you away from how the parser really works under the hood. To me, tags are for when you have a property named Foo that needs to read something called "bar" in JSON.

https://go.dev/play/p/bL6IVsLIWVV

Having a bson tag doesn't mean anything. I've seen plenty of things that people put tags on that never get written to the database. Only the code can tell me if it gets written or not.

Reply:
Yep, the Go JSON parser will accept either `Token` or `token` when decoding JSON, but our OpenAPI spec shows it as `token`, yes? This just ensures we are enforcing that at the code level. 

I know that a general policy is to be "loose" on input and "strict" on output, but we've been burned so many times on the "loose" on input with our device data. The problem is that you either have to normalize everything on input so that on output it is strict (so that clients receiving that output get something consistent) or just be more strict on input. After much time spent trying to clean up "loose" input after-the-fact, I prefer just enforcing strict validation on input.

Regarding the bson tag, I agree with you. There is no guarantee that it is written to Mongo or not, but it does shows that it was designed for writing to Mongo. Sure, not that helpful, but it is something.

Reply:
> This just ensures we are enforcing that at the code level.

But that isn't true. You can't enforce it with the Go JSON lib. "Token" is still accepted. What have you enforced? You've just given yourself a false idea of what will be accepted.

> Sure, not that helpful, but it is something.

As useful as a comment? ;P

Reply:
Mongo will default to all lowercase attributes if there's no bson tag - if you were to add a new attribute that has multiple words you'd have to add a `bson` tag to make it use camel case for consistency with the other database models. It's much easier to miss this if you don't have any bson tags, even if they are not needed now. So I agree with Darin here that best to have both tags for consistency.